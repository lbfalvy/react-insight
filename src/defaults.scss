.insight__undefined, .insight__null {
    color: #0af;
    /*color: #22f;
    text-shadow: 0 0 4px #f44;*/
}
.insight__number { color: #aea; }
.insight__bigint { color: #8d8; }
.insight__boolean { color: #6247fa; }
.insight__string { color: #fb9565; }