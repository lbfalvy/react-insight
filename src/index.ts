export {
    // Config
    Renderer, Describe, Entry, TypemapConfig, merge,
    // Typemap
    Typemap, compileTypemap,
} from './Typemap'
export { describe } from './resolve'
export { Render } from './Render'
export { useTypemap, Provider, Extender } from './context'
export { config, typemap } from './defaults'
export { arrayEntry } from './renderers/Array'
export { asyncIterableEntry } from './renderers/AsyncIterable'
export { errorEntry } from './renderers/Error'
export { functionEntry } from './renderers/Function'
export { objectEntry } from './renderers/Object'
export { promiseEntry } from './renderers/Promise'
export { symbolEntry } from './renderers/Symbol'