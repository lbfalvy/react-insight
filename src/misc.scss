:root {
    background-color: #181818;
    color: #eee;
    font-family: 'Source Code Pro';
}

.insight__input {
    border: none;
    font-family: inherit;
    outline: none;
    color: #eee;
    background: none;
}

.insight__button {
    background: none;
    color: #999;
    &:hover { color: #eee; }
    &:active { color: #444; }
    transition: color .1s linear;
    // font-family: inherit;
    text-transform: uppercase;
    font-weight: 400;
    transform-origin: 50% 100%;
    transform: scale(.8);
    border: none;
    outline: none;
    cursor: pointer;
}