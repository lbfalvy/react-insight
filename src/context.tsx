import React from "react";
import { config } from "./defaults";
import { compileTypemap, merge, Typemap, TypemapConfig } from "./Typemap";

const ctx = React.createContext<TypemapConfig>(config)

export function useTypemap(): Typemap {
    const context = React.useContext(ctx)
    return React.useMemo(() => compileTypemap(context), [context])
}

type ProviderProps = Parameters<React.Provider<TypemapConfig>>[0]

function makeDepArray(cfg: TypemapConfig) {
    return [Object.entries(cfg.typeof ?? {}), cfg.instanceof, cfg.test].flat(Infinity)
}

export function Provider(
    { value, children }: ProviderProps
): React.ReactElement {
    const old = React.useContext(ctx)
    const current = React.useMemo(
        () => merge(old, value),
        [old, ...makeDepArray(value)]
    )
    return <ctx.Provider value={current}>
        {children}
    </ctx.Provider>
}

interface ExtenderProps {
    children: React.ReactNode
    callback: (conf: TypemapConfig) => TypemapConfig
}

export function Extender({ callback, children }: ExtenderProps): React.ReactElement {
    const old = React.useContext(ctx)
    const current = React.useMemo(() => callback(old), [callback, old])
    const preserved = React.useMemo(() => current, makeDepArray(current))
    return <ctx.Provider value={preserved}>
        {children}
    </ctx.Provider>
}