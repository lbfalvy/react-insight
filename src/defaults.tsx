import React from "react"
import { truncate } from "./truncate"
import { compileTypemap, TypemapConfig } from "./Typemap"
import { arrayEntry } from "./renderers/Array";
import { objectEntry } from "./renderers/Object";
import { functionEntry } from "./renderers/Function";
import './defaults.scss';
import { symbolEntry } from "./renderers/Symbol";
import { promiseEntry } from "./renderers/Promise";
import { asyncIterableEntry } from "./renderers/AsyncIterable";
import { errorEntry } from "./renderers/Error";
import { mapEntry } from "./renderers/Map";
import { setEntry } from "./renderers/Set";
import { dateEntry } from "./renderers/Date";

export const config: TypemapConfig = {
    typeof: {
        number: [
            ({ data }) =>
            <span className='insight__number'>
                {data.toString()}
            </span>,
            (n: number) => n.toString()
        ],
        bigint: [
            ({ data }) =>
            <span className='insight__bigint'>
                {data.toString()}n
            </span>,
            (bi, _, length) => truncate(bi.toString(), length-1, true)+'n'
        ],
        boolean: [
            ({ data }) =>
            <span className='insight__boolean'>
                {data.toString()}
            </span>,
            (b: boolean) => b.toString()
        ],
        function: functionEntry,
        string: [
            ({ data }) => 
            <span className='insight__string'>
                "{data}"
            </span>,
            (data, _, length) => `"${truncate(data, length - 2)}"`
        ],
        symbol: symbolEntry,
        undefined: [
            () => <span className='insight__undefined' style={{ color: '#0af' }}>
                undefined
            </span>,
            () => 'undefined'
        ],
        object: objectEntry
    },
    instanceof: [
        [Array, arrayEntry],
        [Promise, promiseEntry],
        [Error, errorEntry],
        [Map, mapEntry],
        [Set, setEntry],
        [Date, dateEntry]
    ],
    test: [
        [x => x === null, [
            () => <span className='insight__null'>null</span>,
            () => 'null'
        ]],
        [x => typeof x == 'object' && Symbol.asyncIterator in x, asyncIterableEntry]
    ]
}

export const typemap = compileTypemap(config)