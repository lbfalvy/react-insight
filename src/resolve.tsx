import React from "react"
import { typemap } from "./defaults"
import { truncate } from "./truncate"
import { Entry, Typemap } from "./Typemap"

export function resolve<T>(o: any, tm: Typemap): Entry<T> {
    const final: Entry<T> = [undefined, undefined]
    for (const [test, entry] of tm.test) {
        if (!test(o)) continue
        const Render = final[0] ??= entry[0]
        const describe = final[1] ??= entry[1]
        if (Render && describe) return final
    }
    for (const [Type, entry] of tm.instanceof) {
        if (!(o instanceof Type)) continue
        const Render = final[0] ??= entry[0]
        const describe = final[1] ??= entry[1]
        if (Render && describe) return final
    }
    const typeofEntry = tm.typeof?.[typeof o]
    if (typeofEntry) {
        final[0] ??= typeofEntry[0]
        final[1] ??= typeofEntry[1]
    }
    return final
}

interface RenderProps {
    data: any,
    tm: Typemap
}

export function RenderWith({ data, tm }: RenderProps): React.ReactElement {
    const [Component] = resolve(data, tm)
    if (!Component) throw new Error('Missing component. Did you merge the default config?')
    return <Component data={data} tm={tm} />
}

export function describe(d: any, tm: Typemap, length = 50): string {
    const entry = resolve(d, tm)
    const string = entry[1]?.(d, tm, length)
    if (!string) throw new Error('Missing describer. Did you merge the default config?')
    return truncate(string, length)
}