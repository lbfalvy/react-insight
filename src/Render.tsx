import React from "react";
import { RenderWith } from "./resolve";
import { Typemap } from "./Typemap";
import { typemap } from './defaults';

export function Render({ data, tm }: { data: any, tm?: Typemap }): React.ReactElement {
    return <RenderWith data={data} tm={tm ?? typemap} />
}