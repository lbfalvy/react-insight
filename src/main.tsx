import React from "react";
import ReactDOM from "react-dom";
import { Render } from "./Render";
import { openCtx } from "./renderers/utils";

const root = document.createElement('div')
document.body.appendChild(root)

class Player {
    readonly type = 'player'
    friends: Player[] = []
    constructor(readonly roles: string[]) {}
}

class MyIterable implements AsyncIterable<number> {
    last = 0;
    [Symbol.asyncIterator]() {
        const itble = this
        return {
            async next() {
                await new Promise(res => setTimeout(res, 1000))
                return {
                    done: 24 <= itble.last,
                    value: itble.last++
                }
            }
        }
    }
}

const asyncIterable = new MyIterable()

function doom() { throw new Error('Something broke') }
function rancid() { doom() }
function getError() {
    try { rancid() } catch(ex) { return ex }
}

const pilot = new Player(['pilot'])
const captain = new Player(['captain'])
pilot.friends.push(captain)
captain.friends.push(pilot)

const map = new Map<any, any>()
map.set('Ronald', 12)
map.set('Peter', 43)
map.set('REEEE', Infinity)
map.set(captain, 'red')

const fibset = new Set<number>([1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144])

function Main(): React.ReactElement {
    return <openCtx.Provider value={false}>
        <div><Render data={undefined} /></div>
        <div><Render data={null} /></div>
        <div><Render data={0.123456789} /></div>
        <div><Render data={123456789123456789n} /></div>
        <div><Render data={new Date()} /></div>
        <div><Render data={true} /></div>
        <div><Render data={"It is proven that there cannot exist an algorithm that proves everything provable, so no matter how good our computers become we will never run out of riddles."} /></div>
        <div><Render data={Symbol('Intrinsically unique, a Symbol can identify a unit of code.')} /></div>
        <div><Render data={function foo(a: any, b: any, c: any) {}} /></div>
        <div><Render data={[(a: any, b: any, c: any) => {}][0]} /></div>
        <div><Render data={new Array(10000).fill(1).map(() => Math.random())} /></div>
        <div><Render data={new Player(['impostor', 'pilot'])} /></div>
        <div><Render data={Promise.resolve('Delivered, as promised')} /></div>
        <div><Render data={Promise.reject(new Error('Something broke'))} /></div>
        <div><Render data={new Promise(() => {})} /></div>
        <div><Render data={getError()} /></div>
        <div><Render data={map} /></div>
        <div><Render data={captain} /></div>
        <div><Render data={pilot} /></div>
        <div><Render data={asyncIterable} /></div>
        <div><Render data={fibset} /></div>
    </openCtx.Provider>
}
ReactDOM.render(<Main />, root)