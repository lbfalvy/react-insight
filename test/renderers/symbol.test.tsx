import React from "react"
import { create } from "react-test-renderer"
import { typemap } from "../../src/defaults"
import { Render } from "../../src/Render"
import { describe as getDescription } from '../../src/resolve'

describe('symbol renderer', () => {
    test('symbol', () => {
        const tpl = Symbol('test symbol')
        const component = create(<Render data={tpl} />)
        expect(component.toJSON()).toMatchSnapshot()
        expect(getDescription(tpl, typemap)).toEqual('Symbol("test symbol")')
    })
})