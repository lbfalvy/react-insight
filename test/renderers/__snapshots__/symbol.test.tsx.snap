// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`symbol renderer symbol 1`] = `
<span
  className="dt-symbol"
>
  <span
    className="symbol_symbol"
  >
    Symbol
  </span>
  <span
    className="symbol_lparen"
  >
    (
  </span>
  <span
    className="symbol_description"
  >
    <span
      className="dt-string"
    >
      "
      test symbol
      "
    </span>
  </span>
  <span
    className="symbol_rparen"
  >
    )
  </span>
</span>
`;
